speechbrain/lobes/models/wav2vec.py

```diff
@@ -0,0 +1,610 @@
+        # when finetuning not but one still wants to mask to regularize
+        if is_pretraining or self.training:
+            if mask is None:
+                feats, mask = apply_mask(  # TODO: refactor
+                    feats,
+                    self.mask_emb,
+                    padding_mask,
+                    mask_prob,
+                    mask_length,
+                    mask_channel_prob,
+                    mask_channel_length,
+                )
+            else:
+                with torch.cuda.amp.autocast(enabled=False):
+                    feats[mask] = self.mask_emb
+
+            feats = feats + self.positional_encoding(feats)
+            feats, _ = self.latent_encoder(feats, padding_mask)
+
+            if is_pretraining and mask is not None:
+                feats = feats[mask].view(feats.size(0), -1, feats.size(2))
+                targets = targets[mask].view(targets.size(0), -1, targets.size(2))
+
+                num_masked = mask.sum()
+                num_samples = mask.numel()
+                results["num_masked"] = num_masked
+                results["ratio_masked"] = num_masked / num_samples
+
+        else:
+            feats = feats + self.positional_encoding(feats)
+            feats, _ = self.latent_encoder(feats, padding_mask)
+
+        if do_final_projection:
+            feats = self.feats_projector(feats)
+
+        if normalize_output:
+            feats = F.layer_norm(feats, feats.size(-1))
+
+        if self.quantiser is None:
+            targets = self.target_projector(targets)
+        else:
+            x = self.quantiser(targets)
+            targets = self.target_projector(x["x"])
+            code_perplex = x["code_perplexity"]
+            prob_perplex = x["prob_perplex"]
+            num_vars = x["num_vars"]
+            temp = x["temp"]
+            diversity_loss = (
+                        num_vars - prob_perplex
+                    ) / num_vars
+            results.update(
+                {
+                    "diversity_loss": diversity_loss,
+                    "code_perplex": code_perplex,
+                    "prob_perplex": prob_perplex,
+                    "num_vars": num_vars,
+                    "temp": temp,
+                }
+            )
+        results["targets"] = targets
+        results["embeddings"] = feats
+        results["latent_l2"] = latent_l2
+        return results
+
+    def freeze(self):
+        for param in self.parameters():
+            param.requires_grad = False
+
+    def unfreeze(self):
+        for param in self.parameters():
+            param.requires_grad = True
+
+    def freeze_latent_extractor(self):
+        for param in self.latent_extractor.parameters():
+            param.requires_grad = False
+
+    def unfreeze_latent_extractor(self):
+        for param in self.latent_extractor.parameters():
+            param.requires_grad = True
+
+    @contextmanager
+    def no_sync(self):
```

Comment:
?

Reply:
when wrapped in DDP this method exists. added this method so that when I call it I don't have to do an awkward check for if the module is DDP or not